### PR TITLE
Multiple package dependencies aka quilt.yml

### PR DESCRIPTION
**Pull request by cosmic-byte**
## Description

Similar to requirements.txt, users need a way to specify that a project depends on one or more package instances:
quilt3 install @quilt.yml
see [here](https://docs.google.com/document/d/1Ev34T-5PCuwuSSXl8CBp_MT7gX13Up-9XRd_LFwDYVM/edit#heading=h.yzmthnk2y6se)

## TODO

- [x] Unit Tests
- [x] Documentation

### CODE REVIEW

File: api/python/quilt3/util.py [hunk 1]
```diff
@@ -534,3 +534,80 @@ def catalog_package_url(catalog_url, bucket, package_name, package_timestamp="la
...
+    return ValidatedPackage(
+        name=package_name,
+        path=sub_package,
+        top_hash=top_hash
+    )
+
+
+def parse_packages(packages: list) -> list:
+    """Validates a list of package names and parse each package to a namedtuple of 'name, path and top_hash'."""
+    parsed = list()
```
Comment (sir-sigurd) [inline api/python/quilt3/util.py:563]:
```suggestion
    parsed = []
```

File: api/python/quilt3/util.py [hunk 2]
```diff
@@ -534,3 +534,80 @@ def catalog_package_url(catalog_url, bucket, package_name, package_timestamp="la
...
+            if name.startswith('@'):
+                name = name[1:]
+            path = pathlib.Path(name)
+            if not path.exists():
+                raise QuiltException(f'{name!r} does not exist in path.')
+            self.parsed_yaml = read_yaml(path)
+            self.from_yaml_file = True
+
+    def get_registries(self):
+        parsed = list()
```
Comment (sir-sigurd) [inline api/python/quilt3/util.py:596]:
```suggestion
        parsed = []
```

File: api/python/quilt3/util.py [hunk 3]
```diff
@@ -534,3 +534,80 @@ def catalog_package_url(catalog_url, bucket, package_name, package_timestamp="la
...
+        except QuiltException:
+            if extract_file_extension(name) not in ('yml', 'yaml'):
+                raise QuiltException(f'{name!r} is not a valid dependency file.')
+            if name.startswith('@'):
+                name = name[1:]
+            path = pathlib.Path(name)
+            if not path.exists():
+                raise QuiltException(f'{name!r} does not exist in path.')
+            self.parsed_yaml = read_yaml(path)
+            self.from_yaml_file = True
```
Comment (sir-sigurd) [inline api/python/quilt3/util.py:593]: `from_yaml_file` looks fully derived from `parsed_yaml`. Do we need separate attribute for it?

File: api/python/tests/test_util.py [hunk 4]
```diff
@@ -67,3 +72,58 @@ def test_validate_url():
...
+    assert parsed.path == expected_sub_path
+
+
+@pytest.mark.parametrize(
+    'name, extension, expected_quilt_version, expected_registries',
+    [
+        ('@quilt.yml', 'yml', 1.0, {'s3://some-bucket': 2, 's3://another-bucket': 3}),
+        ('quilt.yml', 'yml', 1.0, {'s3://some-bucket': 2, 's3://another-bucket': 3}),
+        ('quilt.yaml', 'yaml', 1.0, {'s3://some-bucket': 2, 's3://another-bucket': 3}),
+        ('akarve/cord19', None, None, dict())
```
Comment (sir-sigurd) [inline api/python/tests/test_util.py:101]:
```suggestion
        ('akarve/cord19', None, None, {})
```

File: api/python/tests/test_cli.py [hunk 5]
```diff
@@ -43,3 +47,41 @@ def test_push(self):
...
+
+    with mock.patch('quilt3.Package.install') as mocked_package_install:
+        if registry:
+            main.main(('install', name, '--registry', registry))
+        else:
+            main.main(('install', 'quilt.yml'))
+        mocked_package_install.assert_has_calls(expected_calls)
+        assert mocked_package_install.call_count == expected_no_of_calls
+
+    os.remove("quilt.yml")
```
Comment (sir-sigurd) [inline api/python/tests/test_cli.py:87]: We should clean this up even if test fails and I think that we already isolate tests in a temp dir here:
https://github.com/quiltdata/quilt/blob/a5945a111a3065ecd23e64d069aa67e42492c5f2/api/python/tests/conftest.py#L70-L80

**Comment by sir-sigurd** [comment]
@cosmic-byte, could you rebase this please? (don't forget to [isort](https://pycqa.github.io/isort/#using-isort) before push)

File: api/python/tests/test_util.py [hunk 6]
```diff
@@ -64,3 +67,58 @@ def test_validate_url():
...
+    registries = dict(parser.get_registries())
+
+    assert parser.get_yaml_version() == expected_quilt_version
+    assert registries.keys() == expected_registries.keys()
+
+    for key, value in expected_registries.items():
+        assert len(registries[key]) == value
+
+    # remove created file
+    os.remove(f"quilt.{extension}")
```
Comment (sir-sigurd) [inline api/python/tests/test_util.py:112]: Do we need this?

Comment (cosmic-byte) [reply api/python/tests/test_util.py:112]: No we don't, removing...

File: api/python/tests/test_cli.py [hunk 7]
```diff
@@ -46,6 +48,42 @@ def test_push(self):
...
+                call('akarve/lmnb1', dest=None, dest_registry=None, path=None, registry='s3://another-bucket',
+                     top_hash='c698234'),
+                call('akarve/lmnb3', dest=None, dest_registry=None, path='sub/path', registry='s3://another-bucket',
+                     top_hash='c698234')
+            ],
+            5
+        )
+    ]
+)
+def test_cli_install(name, registry, expected_calls, expected_no_of_calls, set_temporary_working_dir):
```
Comment (sir-sigurd) [inline api/python/tests/test_cli.py:74]: If I'm not missing something `set_temporary_working_dir` is an autouse fixture, so it's not needed here:

```suggestion
def test_cli_install(name, registry, expected_calls, expected_no_of_calls):
```

**Review by akarve — CHANGES_REQUESTED** [review]
Need a full docs section on file syntax with examples.

**Comment by cosmic-byte** [comment]
> Need a full docs section on file syntax with examples.

@akarve  I updated the cli docs with `install` example and syntax.
I guess this is not exactly what you requested, so can you suggest a better way of documenting CLI?

File: api/python/tests/data/quilt.yml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+registries:
+  s3://some-bucket:
+    packages:
+      - asah/gpt3:latest
+      - akarve/lmnb1:c698234
+  s3://another-bucket:
+    packages:
+      - asah/gpt3
+      - akarve/lmnb1:c698234
+      - akarve/lmnb3/sub/path:c698234
```
Comment (sir-sigurd) [inline api/python/tests/data/quilt.yml:11]: @akarve @cosmic-byte 
Intra-package paths (`sub/path` here) are not in spec and I'm not sure if they were requested/needed.

Comment (akarve) [reply api/python/tests/data/quilt.yml:11]: We definitely need and want these. Greg please update the spec.

File: api/python/quilt3/main.py [hunk 9]
```diff
@@ -312,18 +335,39 @@ def create_parser():
...
     install_p.add_argument(
         "name",
-        help=(
-            "Name of package, in the USER/PKG[/PATH] format ([/PATH] is deprecated, use --path parameter instead)"
-        ),
-        type=str,
+        help="Name can be of two representations:\n"
+             "- Name of package, in the USER/PKG[/PATH] format ([/PATH] is deprecated, use --path parameter instead)\n"
+             "- Path to quilt.yaml or quilt.yml package dependency file.\n"
+             "  Syntax:\n"
```
Comment (akarve) [inline api/python/quilt3/main.py:351]: This is a fine start to documenting the file format... but it doesn't belong in CLI help :) It belongs in actual docs in markdown.

File: api/python/tests/data/quilt.yml (hunk 8 above)
Comment (cosmic-byte) [reply api/python/tests/data/quilt.yml:11]: I've updated the spec

File: api/python/quilt3/main.py (hunk 9 above)
Comment (cosmic-byte) [reply api/python/quilt3/main.py:351]: I see

**Review by sir-sigurd — CHANGES_REQUESTED** [review]
Should be tested with invalid `packages` value, i.e. with something like this:
```
version: 1.0
registries:
  s3://bucket1:
    packages:
      asah/gpt1:
      asah/gpt2:
  s3://bucket2:
    packages:
      asah/gpt1: latest
  s3://bucket3:
    packages:
      asah/gpt1:latest
```

File: api/python/tests/data/quilt.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+version: 1.0
```
Comment (sir-sigurd) [inline api/python/tests/data/quilt.yml:1]: I think we should raise exception if it's something different.

**Pull request closed without merging**